libopeniscsiusr/idbm.c

```diff
@@ -0,0 +1,699 @@
+
+#define _rec_str(_key, _recs, _org, _name, _show, _n, _mod) \
+do { \
+	_recs[_n].type = TYPE_STR; \
+	strncpy(_recs[_n].name, _key, NAME_MAXVAL); \
+	if (strlen((char*)_org->_name)) \
+		strncpy((char*)_recs[_n].value, (char*)_org->_name, \
+			VALUE_MAXVAL); \
+	_recs[_n].data = &_org->_name; \
+	_recs[_n].data_len = sizeof(_org->_name); \
+	_recs[_n].visible = _show; \
+	_recs[_n].can_modify = _mod; \
+	_n++; \
+} while(0)
+
+#define _rec_uint8(_key, _recs, _org, _name, _show, _n, _mod) \
+do { \
+	_recs[_n].type = TYPE_UINT8; \
+	strncpy(_recs[_n].name, _key, NAME_MAXVAL); \
+	snprintf(_recs[_n].value, VALUE_MAXVAL, "%" PRIu8, _org->_name); \
+	_recs[_n].data = &_org->_name; \
+	_recs[_n].data_len = sizeof(_org->_name); \
+	_recs[_n].visible = _show; \
+	_recs[_n].can_modify = _mod; \
+	_n++; \
+} while (0)
+
+#define _rec_uint16(_key, _recs, _org, _name, _show, _n, _mod) \
+do { \
+	_recs[_n].type = TYPE_UINT16; \
+	strncpy(_recs[_n].name, _key, NAME_MAXVAL); \
+	snprintf(_recs[_n].value, VALUE_MAXVAL, "%" PRIu16, _org->_name); \
+	_recs[_n].data = &_org->_name; \
+	_recs[_n].data_len = sizeof(_org->_name); \
+	_recs[_n].visible = _show; \
+	_recs[_n].can_modify = _mod; \
+	_n++; \
+} while (0)
+
+#define _rec_uint32(_key, _recs, _org, _name, _show, _n, _mod) \
+do { \
+	_recs[_n].type = TYPE_UINT32; \
+	strncpy(_recs[_n].name, _key, NAME_MAXVAL); \
+	snprintf(_recs[_n].value, VALUE_MAXVAL, "%" PRIu16, _org->_name); \
+	_recs[_n].data = &_org->_name; \
+	_recs[_n].data_len = sizeof(_org->_name); \
+	_recs[_n].visible = _show; \
+	_recs[_n].can_modify = _mod; \
+	_n++; \
+} while (0)
+
+enum modify_mode {
+	_CANNOT_MODIFY,
+	_CAN_MODIFY,
+};
+
+struct idbm_rec {
+	int			type;
+	char			name[NAME_MAXVAL];
+	char			value[VALUE_MAXVAL];
+	void			*data;
+	int			data_len;
+	int			visible;
+	/*
+	 * TODO: make it a enum that can indicate wheter it also requires
+	 * a relogin to pick up if a session is running.
+	 */
+	enum modify_mode	can_modify;
+};
+
+int _idbm_lock(struct iscsi_context *ctx)
+{
+	int fd, i, ret;
+	struct idbm *db = NULL;
+	char strerr_buff[_STRERR_BUFF_LEN];
+
+	assert(ctx != NULL);
+
+	db = ctx->db;
+
+	if (db->refs > 0) {
+		db->refs++;
+		return 0;
+	}
+
+	if (access(LOCK_DIR, F_OK) != 0) {
+		if (mkdir(LOCK_DIR, 0660) != 0) {
+			_error(ctx, "Could not open %s: %d %s", LOCK_DIR, errno,
+				_strerror(errno, strerr_buff));
+			return LIBISCSI_ERR_IDBM;
+		}
+	}
+
+	fd = open(LOCK_FILE, O_RDWR | O_CREAT, 0666);
+	if (fd >= 0)
+		close(fd);
+
+	for (i = 0; i < 3000; i++) {
+		ret = link(LOCK_FILE, LOCK_WRITE_FILE);
+		if (ret == 0)
+			break;
+
+		if (errno != EEXIST) {
+			_error(ctx, "Maybe you are not root? "
+			       "Could not lock discovery DB: %s: %d %s",
+			       LOCK_WRITE_FILE, errno,
+			       _strerror(errno, strerr_buff));
+			return LIBISCSI_ERR_IDBM;
+		} else if (i == 0)
+			_debug(ctx, "Waiting for discovery DB lock on %s",
+			       LOCK_WRITE_FILE);
+
+		usleep(10000);
+	}
```

Comment:
So as long as you don't get errno `EEXIST`, eventually you will get the lock as you'll loop 3000 times and then continue on?

Reply:
Thanks. Fixed.